File: 1406-stone-game-iii/1406-stone-game-iii.cpp
class Solution {
    vector<int>* stones;
    vector<int> dp;
    int n;
    int calculate(int i){
        if(i >= n)
            return 0;
        if(dp[i] != INT_MAX)
            return dp[i];
        int ans = INT_MIN;
        ans = max(ans, stones->at(i) - calculate(i+1));
        if(i + 1 < n)
            ans = max(ans, stones->at(i) + stones->at(i+1) - calculate(i+2));
        if(i + 2 < n)
            ans = max(ans, stones->at(i) + stones->at(i+1) + stones->at(i+2) - calculate(i+3));
        return dp[i] = ans;
    }
public:
    string stoneGameIII(vector<int>& stoneValue) {
        n = stoneValue.size();
        stones = &stoneValue;
        dp.resize(n, INT_MAX);
        int ans = calculate(0);
        if(ans > 0)
            return "Alice";
        if(ans < 0)
            return "Bob";
        return "Tie";
    }
};

File: 1406-stone-game-iii/1406-stone-game-iii.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [sys.maxsize] * (n + 1)
        dp[n] = 0
        
        for i in range(n-1, -1, -1):
            dp[i] = stoneValue[i] - dp[i+1]
            if i + 1 < n:
                dp[i] = max(dp[i], stoneValue[i] + stoneValue[i+1] - dp[i+2])
            if i + 2 < n:
                dp[i] = max(dp[i], stoneValue[i] + stoneValue[i+1] + stoneValue[i+2] - dp[i+3])
        
        if dp[0] == 0:
            return "Tie"
        if dp[0] > 0:
            return "Alice"
        return "Bob"